projects/plugins/jetpack/extensions/blocks/dialogue/components/participants-control.js

```diff
@@ -1,59 +1,304 @@
+ * @returns {Function} React component function.
+ */
+export function SpeakerEditControl( {
+	className,
+	label,
+	participant,
+	participants,
+	transcriptRef,
+	onParticipantChange,
+	onUpdate = () => {},
+	onSelect,
+	onAdd,
+	onClean,
+} ) {
+	const [ editingMode, setEditingMode ] = useState( participant ? EDIT_MODE_SELECTING : EDIT_MODE_ADDING );
+
+	// we use a reducer to force re-rendering the SpeakerEditControl,
+	// passing the `reRenderingKey` as property of the component.
+	// It's required when we want to update the options in the autocomplete,
+	// or when we need to hide it.
+	const [ reRenderingKey, triggerRefreshAutocomplete ] = useReducer( speakersControlReducer, 0 );
+
+	function onActionHandler() {
+		switch ( editingMode ) {
+			case EDIT_MODE_ADDING: {
+				triggerRefreshAutocomplete();
+				return onAdd( getTextContent( create( { html: label } ) ) );
+			}
+
+			case EDIT_MODE_EDITING: {
+				return onUpdate( {
+					...participant,
+					label,
+				} );
+			}
+		}
+	}
+
+	/**
+	 * Funcion handler when user types participant label.
+	 * It can edit a new participant, or add a new one,
+	 * dependeing on the previous values.
+	 *
+	 * @param {string} newLabel - New participant label.
+	 * @returns {null} Null
+	 */
+	function onChangeHandler( newLabel ) {
+		// If the new label is empty,
+		// activate autocomplete, and emit onClean(),
+		// to clean the current participant.
+		if ( ! newLabel?.length ) {
+			setEditingMode( EDIT_MODE_ADDING );
+			return onClean();
+		}
+
+		// Always update the participant label (block attribute).
+		onParticipantChange( newLabel );
+
+		const participantByNewLabel = getParticipantByLabel( participants, newLabel );
+
+		// Set editing mode depending on participant label,
+		// and current conversation participant
+		// tied to this Dialogue block.
+		if ( participant ) {
+			if ( participant.label === newLabel ) {
+				setEditingMode( EDIT_MODE_SELECTING );
+			} else {
+				setEditingMode( EDIT_MODE_EDITING );
+			}
+		} else if ( participantByNewLabel ) {
+			setEditingMode( EDIT_MODE_SELECTING );
+		} else {
+			setEditingMode( EDIT_MODE_ADDING );
+		}
+	}
+
+	// Keep autocomplete options udated.
+	const autocompleter = useMemo( () => {
+		if ( editingMode !== EDIT_MODE_ADDING ) {
+			return [];
+		}
+
+		return [ refreshAutocompleter( participants ) ];
+	}, [ participants, editingMode ] );
+
+	useEffect( () => {
+		setEditingMode( participant ? EDIT_MODE_SELECTING : EDIT_MODE_ADDING );
+	}, [ participant ] );
+
+	return (
+		<DetectOutside
+			className={ classNames( className, {
+				'has-bold-style': label?.length,
+				'is-adding-participant': editingMode === EDIT_MODE_ADDING,
+				'is-editing-participant': editingMode === EDIT_MODE_EDITING,
+				'is-selecting-participant': editingMode === EDIT_MODE_SELECTING,
+			} ) }
+			onFocusOutside={ onActionHandler }
+		>
+			<RichText
+				key={ `re-render-key${ reRenderingKey }` }
```

Comment:
let me take a look. Could you help me to reproduce the issue? I'm forcing a rendering to update the autocomplete, and it should be done only when the `participants` object changes.

Reply:
Yeah, it's a bit sporadic. I see you've removed the reducer, so I'll give it another try and see if I can get a reliable test case.

Reply:
removed here https://github.com/Automattic/jetpack/pull/18646/commits/2cdcf2a771a61bdb400ffa89a5c6eeef7be312bf

I had to add it because the speaker's list did not update once a new one was added. I've tested again, removing this technique and it seems to work. properly. 👍

Reply:
It's difficult to tell for sure, but it seems like I can't reproduce the focus problems I was seeing. I think you're right that this might be a more general problem with the editor, but I think for the context of this block we can consider it solved!